core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -1324,7 +1401,59 @@ class ConnectionQuotas(config: KafkaConfig, time: Time, metrics: Metrics) extend
   private[network] def updateBrokerMaxConnectionRate(maxConnectionRate: Int): Unit = {
     // if there is a connection waiting on the rate throttle delay, we will let it wait the original delay even if
     // the rate limit increases, because it is just one connection per listener and the code is simpler that way
-    updateConnectionRateQuota(maxConnectionRate)
+    updateConnectionRateQuota(maxConnectionRate, BrokerQuotaEntity)
+  }
+
+  /**
+   * Update the connection rate quota for a given IP and updates quota configs for updated IPs.
+   * If an IP is given, metric config will be updated only for the given IP, otherwise
+   * all metric configs will be checked and updated if required.
+   *
+   * @param ip ip to update or default if None
+   * @param maxConnectionRate new connection rate, or resets entity to default if None
+   */
+  def updateIpConnectionRateQuota(ip: Option[String], maxConnectionRate: Option[Int]): Unit = {
+    def isIpConnectionRateMetric(metricName: MetricName) = {
+      metricName.name == ConnectionRateMetricName &&
+      metricName.group == MetricsGroup &&
+      metricName.tags.containsKey(IpMetricTag)
+    }
+
+    def shouldUpdateQuota(metric: KafkaMetric, quotaLimit: Int) = {
+      quotaLimit != metric.config.quota.bound
+    }
+    counts.synchronized {
```

Comment:
yeah, it's not desirable behavior. However, this is more or less in line with
`ClientQuotaManager.updateQuota`, which write locks when updating quota, and also has to iterate over all metrics for any default quota updates.

Reply:
I'm thinking on this a bit more, and I think that only locking on updates to `defaultConnectionRatePerIp` should be sufficient for correctness.

ZK dynamic config changes are processed within one thread, so we will only have one thread executing in `updateIpConnectionRateQuota`. If we want to be really careful about this, we can have `IpConfigHandler` synchronize on `ConnectionQuotas`.

The case we want to avoid by synchronizing on `counts` is that another thread `defaultConnectionRatePerIp` as connection rate limit `A` while calling `inc()`, then we update connection rate and quota metrics to `B`, then the other thread creates a sensor/metric with quota limit `A` => inconsistency.

If we synchronize on `counts` for only updates to `connectionRateForIp/defaultConnectionRate`, we know that any thread that has read a connection rate quota as `A` will finish creating quota metrics with limit `A` before we acquire the `counts` lock.

After acquiring the `counts` lock and updating `connectionRateForIp/defaultConnectionRate` to `B`, any subsequent threads calling `inc()` will read the quota as `B` and create a metric as `B`. We will then be able to update any quota metrics from `A` to `B`, knowing that all new ones will be created as `B`.

The only issue remaining is that we can get concurrent reads of `connectionRatePerIp`, but we can just replace `mutable.Map` with `ConcurrentHashMap`.

Reply:
Thanks for the explanation, David. I agree with your approach. In the end, we want to ensure that we don't create a Sensor/Metrics while `connectionRateForIp/defaultConnectionRate` are updated. It think that it is fine to rely `counts` lock to achieve this but we should add some comments to make this clear. I was considering if we should use a different and more explicit lock to achieve this but I am fine with using `counts` as long as we add good comments.

Having `updateIpConnectionRateQuota` synchronized on `ConnectionQuotas` sounds good to me. I think that we have to ensure that metrics are updated by one thread only. Otherwise, we may have inconsistencies.

Reply:
I reviewed the commit with the locking change, and it looks correct to me. Thanks for writing up a thorough evaluation of the new logic. I think using `counts` as a lock is reasonable here, because we need to synchronize a lot of the same data that is used on the `inc()` path (protected by `counts`).  I agree about adding clear comments.